Fix dagger implementation for matrices of operators and constants.

hodgestar opened the pull request:
#### References to other Issues or PRs

Fixes #16959

#### Brief description of what is fixed or changed

The implementation of adjoint for matrices did not take into account the case where elements might be non-scalars. This PR corrects the implementation to take the *adjoint* of each element rather than the *conjugate* of each element.

The dagger operator failed for constant scalar values (e.g. `5`, `1 + 1j`).

The dagger module monkeypatched the class name of the adjoint operator and its repr. These were removed for sanity.

#### Other comments

Tests were added for applying the Dagger operator to constants and matrices of operators.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * Matrix .adjoint() and .H now return the correct result when the elements of the matrix are not scalars (before it was assumed that the adjoint of each element was always its conjugate, which is not true for matrices of matrices or matrices of operators).

* physics.quantum
  * Dagger(x) now returns conjugate(x) when x is a constant (before it raised an error).
  * Dagger(x) now returns the correct result when x is a matrix of operators (before it returned the conjugate of the contained operators instead of the adjoint).
  * Importing the quantum module no longer changes the name of the adjoint class to Dagger.
<!-- END RELEASE NOTES -->

hodgestar committed:
Add tests for daggers of constants and matrices of operators.

hodgestar committed:
Remove monkeypatching of adjoint.__name__ and fix incorrect call to base class __new__.

hodgestar committed:
Fix implementation of adjoint for matrices with non-scalar elements.

oscarbenjamin commented:
Thanks for working on this. Looks like there are some test failures...

Does this also change the output of `mat.H` or is that correct as it is?

Either way can you add a test for `mat.H` next to the one for `Dagger(mat)` so we can the similarity/difference?

The release notes are unclear. Can you explain in terms relevant for a user e.g.: Dagger(x) now gives conjugate(x) for scalar x.

What does the note about monkeypatching refer to?

oscarbenjamin commented:
It would also be good to compare (in the tests) what mat.transpose, mat.conjugate, etc do. Let's make sure we think that these are all doing the right thing and then test so they don't break in future.

asmeurer commented on sympy/physics/quantum/tests/test_dagger.py:
```diff
@@ -1,10 +1,17 @@
+def test_constants():
+    assert Dagger(0) == 0
+    assert Dagger(1) == 1
+    assert Dagger(1 + 1j) == 1 - 1j
```
```suggestion
    assert Dagger(1 + I) == 1 - I
```

hodgestar committed:
Use I instead of j to avoid unnecessary use of floating point numbers.

hodgestar commented:
> Thanks for working on this. Looks like there are some test failures...

Yep, will dig into.

> Does this also change the output of `mat.H` or is that correct as it is?

It also changes the output of mat.H.

> Either way can you add a test for `mat.H` next to the one for `Dagger(mat)` so we can the similarity/difference?

Will do! I need to add tests for the matrix code I changed.

> The release notes are unclear. Can you explain in terms relevant for a user e.g.: Dagger(x) now gives conjugate(x) for scalar x.

Updated!

> What does the note about monkeypatching refer to?

The monkeypatching that was removed was at https://github.com/sympy/sympy/pull/17189/files#diff-468ef3d693d393047d38a18dd41e7cceL88 -- I think the idea was to make the repr of generic adjoint look more familiar to physicists, but the cost of monkeypatching another class when the dagger module is imported doesn't seem worth it.

hodgestar committed:
Remove unnecessary spaces.

hodgestar committed:
Add tests for taking the adjoint and hermitian conjugate of a matrix of matrices.

oscarbenjamin commented on sympy/matrices/tests/test_commonmatrix.py:
```diff
@@ -510,6 +516,12 @@ def test_conjugate():
+def test_hermitian_conjugate_with_matrix_elements():
+    submat = Matrix([[0, I], [1, 0]])
+    mat = Matrix([[0, submat], [1, 0]])
+    assert mat.H == mat.adjoint()
```
Since this test function is very similar to the one above I would merge them into one function. Also I would prefer the rhs of these asserts to be as explicit as possible.

You could have a bunch of
```
assert mat.H ==
assert mat.transpose() == 
assert mat.conjugate().transpose() ==
...
```
(same goes for matrix of operators)

These tests are cheap and I think it's good to show all possibilities right next to each other so it's clear to anyone looking at the test code exactly how/why they are similar/different and that any differences are intentional.

hodgestar commented:
@oscarbenjamin I've discovered that the following modules make similar assumptions that `adjoint = conjugate transpose`:
* sympy.functions.elementary.complexes
* sympy.matrices.expressions.adjoint
* sympy.matrices.expressions.transpose
* sympy.tensor.array.ndim_array

So I also need to think about how to address those without breaking everything. Are there any of those that should keep the assumption? Any thoughts on how to keep some of the inferences that are currently allowed?

oscarbenjamin commented:
Can you point to an example of where some code makes this assumption and explain why it isn't valid? Maybe this just needs to be fixed in more places.

hodgestar commented:
> Can you point to an example of where some code makes this assumption and explain why it isn't valid? Maybe this just needs to be fixed in more places.

E.g. https://github.com/sympy/sympy/blob/master/sympy/matrices/expressions/adjoint.py#L56-L57 -- `Adjoint._eval_conjugate` assumes that the conjugate of the adjoint is the transpose, which would not be true for a matrix of operators.

oscarbenjamin commented:
> `Adjoint._eval_conjugate` assumes that the conjugate of the adjoint is the transpose, which would not be true for a matrix of operators.

I guess that to do any manipulation of matrix symbols we will need to have some assumption about what is in the matrix. Ideally you would specify the field/ring over which the matrix is defined.

If we say that a MatrixSymbol only represents a matrix of complex numbers then this is correct right? Is this example causing problems for this PR?

hodgestar commented:
> I guess that to do any manipulation of matrix symbols we will need to have some assumption about what is in the matrix. Ideally you would specify the field/ring over which the matrix is defined.

Correct.
 
> If we say that a MatrixSymbol only represents a matrix of complex numbers then this is correct right? Is this example causing problems for this PR?

Yes, that would be correct. If this PR is fixing #16959, then I think yes it is causing problems for this PR because `transpose(adjoint(Matrix([[A, 0], [0, B]])))` would give `Matrix([[conjugate(A), 0], [0, conjugate(B)]])` which would be wrong unless `conjugate(A) == adjoint(A)` and `conjugate(B) == adjoint(B)`.

oscarbenjamin commented:
Maybe these other things will need to be changed as well. The fix for this might need to go deeper than it initially seemed...

Don't be afraid to change things that need to be fixed or to change tests that are not ultimately correct.

jksuom commented:
Perhaps the names could be organized according to the following guidelines.

- `conjugate`: Scalar method (though can be applied element-wise to matrices).

- `transpose`: Matrix method (that does not change the elements). No-op for scalars.

- `adjoint`: Same as conjugate for scalars, combination of adjoint and transpose in general.

oscarbenjamin commented:
> Perhaps the names could be

Seems reasonable to me.

Should mat.H be the conjugate transpose rather than the adjoint then?

asmeurer commented:
Is the is_scalar attribute on Expr the correct way to check if something is a scalar in this context?

hodgestar commented:
Is there a way to check on the matrix class what kind / field its elements
are supposed to be from? Is that something one should add? Might be a bit
painful for users to be that explicit?

oscarbenjamin commented:
> Is there a way to check on the matrix class what kind / field its elements are supposed to be from?

Currently no. I think that ultimately that would be a good thing to have but that's a big task to implement. Ideally both MatrixSymbol and ImmutableMatrix instances would have some sort of associated ring like a Poly does. It wouldn't need to be painful for users: ImmutableMatrix can default to checking the types of the arguments and MatrixSymbol can default to complex numbers. The example here would need a ring of operators to be defined somehow though...

I don't think that is needed here though is it?

asmeurer commented:
Some discussion on that is at https://github.com/sympy/sympy/issues/10265

hodgestar commented:
Haven't forgotten about this -- just traveling at the moment.

oscarbenjamin commented:
> Haven't forgotten about this -- just traveling at the moment.

That's obviously fine :). Are you planning to pick this up again though?

(If not if can be left for someone else to pick up at some point)

hodgestar commented:
At the moment I'm not likely to get back to this -- it ended up being a bigger chunk of work than I was expecting. Anyone else please feel free to take over. Also happy for this to be closed and picked up some other time.

czgdp1807 commented:
@hodgestar Would you still like to work on this? Please do not close this PR.

hodgestar commented:
@czgdp1807 I'm still not likely to get back to this, but happy for someone else to take it over if they like.

ellisonbg reviewed (approved):
There was a minor comments about consolidating the tests, but I probably not something that should block merging. Approving.

ellisonbg commented:
Will also need a rebase.

oscarbenjamin commented:
This will need to be changed so that the code is added in `matrixbase.py` rather than `common.py`. The various matrix classes have been consolidated.

oscarbenjamin commented:
https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-matrix-mixin-classes

ellisonbg commented:
Thanks @oscarbenjamin

akhmerov commented:
I have done this now in #27422

akhmerov commented:
This should be closed because it's superceded by #27422
